### PR TITLE
(Closes #426) Add convert argument to open().

### LINKED ISSUE #426: Add support for the open convert specifier (which is not part of the standard)
fparser fails to parse:
```
       open(unit       = MyUnit,       &
            file       = trim(filename), &
            form       = 'UNFORMATTED',  &
            access     = 'SEQUENTIAL',   &
            action     = 'WRITE',        &
            convert    = 'BIG_ENDIAN',   &
            iostat     = E_IO)
```

I believe this is because the "convert" is not part of the Fortran standard but it is supported by gfortran, nvfortran and ifort
https://gcc.gnu.org/onlinedocs/gfortran/CONVERT-specifier.html

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
In the interests of speed (of implementation) I haven't added full testing of the OPEN statement here.
Ready for review from either @sergisiso or @rupertford.

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -8615,6 +8619,7 @@ def match(string):
                     "ACTION",
                     "ASYNCHRONOUS",
                     "BLANK",
+                    "CONVERT",
```
Comment (rupertford) [inline src/fparser/two/Fortran2003.py:8622]: Could you please add CONVERT as an optional extension as we do for the other extensions (see utils.py)

**Review by rupertford — CHANGES_REQUESTED** [review]
Some additional tests, refactoring and making the extension optional are required - please see inline comments.

File: src/fparser/two/Fortran2003.py (hunk 1 above)
Comment (arporter) [reply src/fparser/two/Fortran2003.py:8622]: Oh yes, I'd forgotten about that. Apologies.

File: doc/source/fparser2.rst [hunk 2]
```diff
@@ -353,6 +353,19 @@ extension adds support for the dollar descriptor in fparser.
 For more information see
 https://software.intel.com/en-us/fortran-compiler-developer-guide-and-reference-dollar-sign-and-backslash-editing
 
+CONVERT argument to OPEN
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/source/fparser2.rst:356]: Done.

File: src/fparser/two/tests/fortran2003/test_openstmt_r904.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+"""
+Module containing py.test tests for the Fortran2003 OPEN statement (R904).
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_openstmt_r904.py:67]: Oh yes. Split and improved.

**Comment by arporter** [comment]
Ready for another look now. Ran across #432 while trying to check the documentation. Happily that's a separate issue though :-)

**Comment by arporter** [comment]
Taking this back as I've discovered it doesn't work if one requests a F2008 parser instead of F2003!

**Comment by arporter** [comment]
This has got a bit bigger as, on looking at the Fortran2008 implementation of Open (which I missed first time around), I realised that there was a lot of code duplication that could be factored out. I also slightly refactored the way that we handle the list of supported extensions (we now have a method that returns the list rather than just a module-scope variable containing a list). This makes testing a bit cleaner as monkeypatch can be used more easily.

Ready for another look now.

**Review by sergisiso — APPROVED** [review]
@arporter Again, I gave a quick read to this PR since this was already reviewed once to check that I agree with the general solution and that the previously raised comments from @rupertford are now all resolved. This is ready to merge.

**Pull request merged**